File: leetcode/link_list/0092_Reverse_Linked_List_II/main.py
#!/bin/env python3
#coding:utf-8

class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

    def Print(self):
        curr = self
        s = ""
        while curr:
            s += "%d->" % (curr.val)
            curr = curr.next
        print(s) 


class Solution:
    def reverseBetween(self, head, m, n):
        dummy = ListNode(0)
        dummy.next = head
        Head = dummy
        for i in range(m - 1):
            Head = Head.next
        pre = Head.next
        node = pre.next
        while node and m < n:
            pre.next = node.next
            node.next = Head.next
            Head.next = node
            node = pre.next
            m += 1
        return dummy.next

if __name__ == '__main__':
    a = ListNode(1)
    b = ListNode(2)
    c = ListNode(3)
    d = ListNode(4)
    e = ListNode(5)

    a.next = b;
    b.next = c; 
    c.next = d; 
    d.next = e

    a.Print()
    Solution().reverseBetween(a, 2, 4).Print()

File: leetcode/link_list/0092_Reverse_Linked_List_II/main.cpp
//
// Created by colin on 2018-12-11.
//

//反转从位置 m 到 n 的链表。请使用一趟扫描完成反转。
//
//说明:
//1 ≤ m ≤ n ≤ 链表长度。
//
//示例:
//
//输入: 1->9->2->3->4->5->NULL, m = 2, n = 5
//输出: 1->4->3->2->9->5->NULL

#include "../listnode.h"
#include <iostream>

class Solution {
public:
    //不容易
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if (m >= n) return head;
        int idx = 0;
        ListNode dummy(0);
        dummy.next = head;

        ListNode* s = NULL;
        ListNode* p = NULL;

        ListNode* curr = head;
        ListNode* pre = &dummy;

        while (curr) {
            idx++;
            ListNode* next = curr->next;
            if (idx >= m && idx <=n) {
                if (idx == m) {
                    p = pre;
                    s = curr;
                } else if (idx == n) {
                    curr->next = pre;
                    p->next = curr;
                    s->next = next;
                    break;
                }
                curr->next = pre;
            }
            pre = curr; 
            curr = next;
        }
        return dummy.next;
    }
};

int main() {
    ListNode* l1 = ListNode::MakeList({1, 9, 2, 3, 4, 5});
    ListNode* l2 = ListNode::MakeList({1, 9});
    ListNode::PrintList(Solution().reverseBetween(l1, 2, 5));
    ListNode::PrintList(Solution().reverseBetween(l2, 1, 2));
    return 0;
}